### ISSUE TITLE
filter_select() and friends should be able to respond to crosstalk events

### ISSUE THREAD

**Issue by cpsievert** [rstudio/crosstalk#34]
For instance, [this is how `filter_select()` sets crosstalk events](https://github.com/rstudio/crosstalk/blob/de12a5297fbcf30506b0bfa95b6a10a45ce50ac3/javascript/src/input_selectize.js#L33-L49), but there is no logic for *responding* to crosstalk events (i.e., updating of items)

**Comment by jcheng5** [comment]
Sorry, I don't understand. The two events in Crosstalk are that someone made a selection or that someone changed a filterhandle. Which of those events do you mean by "updating of items"? Neither of those seem to me like they'd affect a `filter_select`, as it isn't affected by selection nor by other filterhandles.

**Comment by cpsievert** [comment]
> The two events in Crosstalk are that someone made a selection or that someone changed a filterhandle.

I get that, but it seems as though you are tying a "selection" events to *direct* manipulation of graphical objects, and "filter" to *indirect* manipulation. I think of it more generally -- either event should be able to be triggered via direct or indirect manipulation. IMO, this is especially important for selection events (if you're curious why, [see this short explanation](https://cpsievert.github.io/plotly_book/linking-views-without-shiny.html#selection-via-indirect-manipulation)), which is why I [implemented](https://github.com/ropensci/plotly/blob/57e9391aaf083ac2e6b9809fefe91593a43cb359/inst/htmlwidgets/plotly.js#L411-L417) a selectize widget that *responds to a crosstalk selection event* (i.e., the selectize items are updated upon a selection event). Notice the bidirectional communication between the widget and the plot in this gif:

![plotlyselectize](https://cloud.githubusercontent.com/assets/1365941/25015877/5455d2be-2043-11e7-89da-22dd4eaf53fa.gif)

> Which of those events do you mean by "updating of items"?

Both...for the reasons I layout above, I would like to see crosstalk widgets update their value when a relevant event is triggered (regardless of *how* they are triggered). I would also like to see crosstalk widgets that can trigger selection events (e.g., `select_checkbox()`, which would be just like `filter_checkbox()`, except triggers a selection event, rather than a filter event).


PS. this is how I view the difference between selection and filter:

* __Filter:__ completely remove the "non-selected" data, and update the layout accordingly.
* __Selection:__ dim the "non-selected" data, keeping the layout fixed.

**Comment by jcheng5** [comment]
I agree that direct vs. indirect is orthogonal to filter and selection. Your P.S. captures the important aspects.

But one other aspect is that there is only a single "selected" state for a given group, and whoever mutates it last completely overrides any previous values; while for filters, each widget or whatever has its own handle that has its own independent state (set of keys), and the group's overall filter state is the intersection of all of those sets.

Therefore, it totally makes sense that you created a selectize widget that both sets and responds to the _selection_ state; whenever a selection event occurs in another widget, just override your current state to reflect the new state. But doing so with a _filter_ control is a different proposition; when another filter component has changed its contribution to the overall filter state, what does that mean for your state?

There is at least one good answer I know of to that hypothetical question, which is, your state doesn't change but you filter down your visible choices to the intersection of every other filter handle's set _except your own_. But I don't think you can implement this with Crosstalk today, and I made that decision consciously.

I get that having filter controls that respond to other filter controls is a reasonable and quite desirable feature, e.g. pretty much every dc.js/crossfilter demo. I punted on this scenario for v1 of crosstalk--not because the crosstalk part of it would be difficult to implement, but because the burden on widget developers to support this correctly seemed unreasonably high (unless you started writing your widget from scratch with this kind of interconnected filtering in mind, as was the case for dc.js).

**Comment by cpsievert** [comment]
Ah, I see, I didn't know that about the automatic filter state resolution, which makes sense. That now makes me wonder: wouldn't it make sense to update a filter's choices based on the current state? 

Also, I agree that a filter's current value shouldn't react to whatever another filter is doing (essentially because of the AND logic -- for details, see below). 

Anyway, I still think there should be some sort way to set a filter's value at print time https://github.com/rstudio/crosstalk/issues/16#issuecomment-293721581 

<details>

> But one other important aspect is that there is only a single "selected" state for a given group, and whoever mutates it last completely overrides any previous values; while for filters, each widget or whatever has its own handle that has its own independent state (set of keys), and the group's overall filter state is the intersection of all of those sets.

What you are describing for the overall filter state sounds like something I'd call *persistent* selection, with *AND* logic, and a *filter* operation. In plotly, I've mostly been thinking about and implementing *transient*/*persistent* selection, with *OR* logic, and a *brush*/*select* operation. Of course, AND is more intuitive for a filter operation, but I'd argue OR is more intuitive for brushing, especially if the goal is to make comparisons. https://vimeo.com/190011823

For brushing, there are good use cases for logic other than OR ([cranvas has this](https://github.com/ggobi/cranvas/blob/master/R/brush.R#L47-L49)), and it's on my [TODO list to implement](https://github.com/ropensci/plotly/issues/836). 

> Therefore, it totally makes sense that you created a selectize widget that both sets and responds to the selection state; whenever a selection event occurs in another widget, just override your current state to reflect the new state. But doing so with a filter control is a different proposition; when another filter component has changed its contribution to the overall filter state, what does that mean for your state?

It depends on the logic for resolving the selection sequence. At this time, I have no issue with *always* using AND logic for filter operations, but selection/brushing is another ball-game.

In fact, I think crosstalk's selection handle should have a more explicit way to set/get the selection mode (i.e., transient/persistent) and the logic (e.g., OR, AND, XOR, COMPLEMENT). At the very least, that would make [this leaflet hack](https://github.com/rstudio/leaflet/pull/346/files#diff-8e0e6e242b99825df0596336e5ad4d7aR877) to respect plotly's persistent mode more official. That's not to say crosstalk should automatically resolve the selection state, though. Mainly just a more official mechanism to track the selection history and logic.
</details>

**Comment by SteveBronder** [comment]
Wanted to bump this, is there currently a way to get a `select_filter`'s choices to depend on another `select_filter`? in pseudocode

```R
dessert_df = data.frame(
  rating = c(1,2,3,4),
  dessert_type = c("Candy", "Pie", "Candy", "Pie"),
  dessert_subtype = c("KitKat", "Lemon", "Snickerdoodle", "Pecan")
)
dessert_sd = SharedData$new(dessert_df)

dessert_types_ft = filter_select("dessert_type", "Dessert Type",
    dessert_sd, ~dessert_type)
dessert_subtypes_ft = filter_select("dessert_subtypes", "Dessert Subtypes",
    dessert_sd, ~dessert_subtype)
```

Where `dessert_subtypes_ft`'s choices will depend on the `dessert_type_ft` choice?

**Comment by riemerra** [comment]
crosstalk::filter_checkbox leaflet is not working when I add a second addCircles to leaflet.  The filter stops working.  See minimal example below.  I apologize in advance is this question is not according to form, I am new to rstudio and this is my first questions.

---
title: "crosstalk::filter_checkbox leaflet"
output: html_document
date: "2022-12-14"
---
```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```
```{r, echo=FALSE}
library(leaflet)
library(crosstalk)

tstd = data.frame(lng=-116.3013+rnorm(10)/100,lat=33.1005+rnorm(10)/100,
           Level1=c("L","L","L","L","L","H","H","H","H","H"))

sd_tstd <- SharedData$new(tstd)

crosstalk::filter_checkbox("Level1", "Level One", sd_tstd, ~Level1, inline = T)
```

```{r, echo=FALSE, out.width = '100%'}
leaflet(sd_tstd) %>% addTiles() %>% setView(-116.3013, 33.1005, zoom=13) %>% 
  addCircles(lng=~lng, lat=~lat, radius=99, stroke=T , weight=1, fill="Grey",  color="black", opacity=1, group = "No Fill" ) %>%   addCircles(lng=~lng, lat=~lat, radius=99, stroke=T , weight=1, fill="Blue",  color="black", opacity=1, group = "With Fill" )
